rocksdb_replicator/replicated_db.cpp

```diff
@@ -141,6 +148,43 @@ RocksDBReplicator::ReplicatedDB::ReplicatedDB(
           + FLAGS_replicator_client_server_timeout_difference_ms));
 }
 
+/*
+ * Helper function to reset the upstream IP after querying for latest leader from helix.
+ */
+void RocksDBReplicator::ReplicatedDB::resetUpstream(std::shared_ptr<ReplicatedDB> db) {
+  if (FLAGS_replicator_zk_cluster.empty() || FLAGS_replicator_helix_cluster.empty()) {
+    LOG(ERROR) << "[resetUpstream] ZK cluster or helix cluster name not provided.";
+    return;
+  }
+
+  // Query helix only FLAGS_replication_error_reset_upstream_percentage %ge of the requests.
+  // This is to avoid excessive load to Helix (zk)
+  if (folly::Random::rand32(0, 99) >= FLAGS_replication_error_reset_upstream_percentage) {
+    LOG(ERROR) << "[resetUpstream] Skip checking latest upstream for " << db->db_name_;
+    return;
+  }
+
+  auto segment_name = common::DbNameToSegment(db->db_name_);
+  auto helix_partition_name = common::DbNameToHelixPartitionName(db->db_name_);
+  LOG(ERROR) << "[resetUpstream] Zookeeper: " << FLAGS_replicator_zk_cluster << " cluster: " <<
+    FLAGS_replicator_helix_cluster << " segment: " << segment_name
+     << " helix_partition_name: " << helix_partition_name;
+  auto leader_id = common::GetLeaderInstanceId(
+    FLAGS_replicator_zk_cluster, FLAGS_replicator_helix_cluster, segment_name, helix_partition_name);
+  LOG(ERROR) << "[resetUpstream] Leader for " << helix_partition_name << " is " << leader_id;
+
+  if (leader_id.length() > 0) {
+    std::vector<std::string> tokens;
+    folly::split("_", leader_id, tokens);
+    if (tokens.size() == 2 && tokens[0].compare(db->upstream_addr_.getAddressStr()) != 0) {
+      incCounter(kReplicatorLeaderReset, 1, db->db_name_);
+      LOG(ERROR) << "[resetUpstream] Resetting upstream for " << db->db_name_ << " to " << tokens[0];
+      uint16_t port = atoi(tokens[1].c_str());
+      db->upstream_addr_.setFromIpPort(tokens[0], port);
```

Comment:
I guess leader_id got from helix contains service port, while replicator uses a separate port for replication.
Sorry, this is a bit confusion.
You may take a look at this code: https://github.com/pinterest/rocksplicator/blob/b70224c1e3f1b8ff7ba1d0b854cacb5efa5b7263/rocksdb_admin/admin_handler.cpp#L281

Reply:
Oh I had missed this, thanks